File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int lb = lowerBound(nums, target);
        int[] result = new int[2];
        if(lb == nums.length || nums[lb] != target){
            result[0] = -1;
            result[1] = -1;
            return result;
        }
        int ub = upperBound(nums, target);
        result[0] = lb;
        result[1] = ub -1;
        return result;
    }

    public static int lowerBound(int[] arr, int x){
        int low = 0;
        int high = arr.length -1;
        int ans = arr.length;
        while(low <= high){
            int mid = low + (high-low)/2;
            if(arr[mid] < x){
                low = mid+1;
            }else{
                ans = mid;
                high = mid -1;
            }
        }
        return ans;
    }

    public static int upperBound(int[] arr, int x){
        int low = 0;
        int high = arr.length -1;
        int ans = arr.length;
        while(low <= high){
            int mid = low + (high-low)/2;
            if(arr[mid] <= x){
                low = mid+1;
            }else{
                ans = mid;
                high = mid -1;
            }
        }
        return ans;
    }
}

File: Leet Code/0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 // binary search approach 
var searchRange = function(nums, target) {
    let result = new Array(1);
    result[0] = (binarySearch(nums, target, true)); // true means search in left
    result[1] = (binarySearch(nums, target, false));
    return result;
};

function binarySearch(arr, p, istarget){
    let low = 0;
    let high = arr.length -1;
    let ans = -1;
    while(low<=high){
        let mid = low + Math.floor((high-low)/2);
        if(arr[mid] < p){
            low = mid+1;
        }else if(arr[mid] > p){
            high = mid-1;
        }else{
            ans = mid;
            if(istarget){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
    }
    return ans;
}
